### DISCUSSION TITLE
Socket slow if data size is above a certain limit

### DISCUSSION THREAD

**Discussion by QuirkyCort** [micropython/micropython#14748] [ESP32]
I've implemented a simple webserver on my ESP32 running micropython (...tested on 1.19, 1.21, and 1.22), and noticed that beyond a certain size, the time taken to transmit the data rises significantly. Test results...

test4.jpg (4357 bytes): ~20ms
test3.jpg (6158 bytes): ~200ms

Despite being only 41% larger, the second file takes about 10 times as long to transmit.

The relevant portion of the code is as follows:

```
def send_file(self, filename):
        self.client_connection.sendall(b'HTTP/1.0 200 OK\r\n\r\n')
        with open(filename, 'rb') as f:
            while True:
                d = f.read(1024)
                if d == b'':
                    break
                self.client_connection.sendall(d)
        self.client_connection.close()
```

Pre-reading the file into a bytes object doesn't improve the transmission time, so the bottleneck isn't in the read().

Timing is done with a simple...

```
start = time.ticks_ms()
ezhttpd.send_file('test3.jpg')
print(time.ticks_ms() - start)
```

The ESP32 is connected to my router in the above test, but I've observed the same issue when running the ESP32 as an AP and connecting directly to it.

Any ideas what I'm doing wrong here?

**Comment by robert-hh** [comment]
Since you read in the loop, the slow-down can as well be caused by the receiver. I have a small script used for testing which just sends. And that one does not slow down with larger data sizes.

**Reply by QuirkyCort** [reply]
I've also tested pre-reading the entire file into a buffer, and sending it all at once with a single sendall() (...no loop), but the results are the same. I timed only the send, so the slow-down isn't caused by the file read.

On the receiving end, I've tried both Chrome and curl with the same results.

Any suggestions on what I can check?

**Comment by QuirkyCort** [comment]
I've further simplified the program (...just sockets with no http), and used a cPython program on the receiving end. The results are still the same; sending a small file (4357 bytes) takes only 2ms everytime, while a slightly larger file (6158 bytes) takes between 30ms to 300ms (...mostly around 200ms to 300ms).

The sender program on the ESP32...

```
import network
import time
import socket

print('started')

file1 = open('test4.jpg', 'rb')
buf = file1.read(-1)
file1.close()

# Connect to WiFi
wifi = network.WLAN(network.STA_IF)
wifi.active(True)
wifi.connect('XXX', 'YYY')
while not wifi.isconnected():
    pass
print('wifi connected')
print(wifi.ifconfig()[0])

s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
s.bind(('', 1337))
s.listen(0)

while True:
    conn, addr = s.accept() # Blocks!
    print('Connection from : ' + addr[0])

    start = time.ticks_ms()
    conn.sendall(buf)
    print(time.ticks_ms() - start)

    conn.close()
```

The receiver program running on cPython (Linux)...

```
import socket

s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
s.connect(('192.168.1.228', 1337))

sum = 0
while True:
    data = s.recv(10240)
    sum += len(data)
    if len(data) == 0:
        break
print(sum)
```

**Comment by robert-hh** [comment]
Actually I see similar figures in my little test program, which just uses `nc` (netcat) as receiver. For a 5k block I get like 2.6 MByte/s. At 6k it collapses to ~600 kByte/s and stays at that level for any larger transfer.
Testing with iperf3, I get data rates of ~2 MBit/s or ~250 kByte/s.
Tested with a generic ESP32. I would assume that this is the behavior of the esp-idf.

**Reply by QuirkyCort** [reply]
That's a much higher rate than what I'm getting; I'm seeing only ~12 kByte/s.

Can you share your test code?

**Reply by robert-hh** [reply]
Sure. The MicroPyhton version of iperf3.py is at https://github.com/micropython/micropython-lib/tree/master/python-ecosys/iperf3. The PC version of Linux should already be installed at your PC.
My test code:
```py
import socket
from time import sleep_us, ticks_ms
import os

def xmit(count, blocks=1, delay=1000):
    blen = blocks * 1024
    b = bytearray(blen)
    for i in range(blen):
        b[i] = os.urandom(1)[0]
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.connect((PC_IP_ADDRESS, 1234))
    mod = count // 100
    if mod < 10:
        mod = 10
    if mod > 10000:
        mod = 10000
    start = ticks_ms()
    sent = 0
    for i in range(count):
        try:
            s.sendall(b)
            sent += blen
            sleep_us(delay)
            if (i % mod) == 0:
                print(i)
        except Exception as e:
            print(i)
            print(e)
            break
    duration = ticks_ms() - start
    s.close()
    print(sent, "bytes sent", duration/count, "ms/frame", count * blen / (duration * 1000), "MB/s")
```
You have to replace PC_IP_ADDRESS by the value at your set-up.
